src/util/ser.rs

```diff
@@ -0,0 +1,373 @@
+use secp256k1::key::PublicKey;
+use bitcoin::blockdata::script::Script;
+use bitcoin::util::hash::Sha256dHash;
+use std::marker::Sized;
+
+pub const MAX_VEC_SIZE: usize = 32 * 1024 * 1024;
+
+pub enum Error {
+	CorruptedData,
+	BadPublicKey,
+	IoError(io::Error),
+	InvalidLength,
+	InvalidValue,
+}
+
+impl fmt::Debug for Error {
+  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		match *self {
+			Error::CorruptedData => f.write_str("corrupted data"),
+			Error::BadPublicKey => f.write_str("Bad public key"),
+			Error::IoError(ref e) => write!(f, "{}", e),
+			Error::InvalidLength => f.write_str("Invalid length"),
+			Error::InvalidValue => f.write_str("Invalid value"),
+		}
+  }
+}
+
+impl From<io::Error> for Error {
+  fn from(e: io::Error) -> Self {
+		Error::IoError(e)
+  }
+}
+
+pub struct Writer<W> { writer: W }
+pub struct Reader<R> { reader: R }
+
+pub trait Writeable<W: Write> {
+	fn write(&self, writer: &mut Writer<W>) -> Result<(), Error>;
+}
+
+pub trait Readable<R>
+	where Self: Sized,
+				R: Read
+{
+	fn read(reader: &mut Reader<R>) -> Result<Self, Error>;
+}
+
+macro_rules! writer_fn {
+	($name:ident, $val_type:ty, $convfn:ident) => {
+		#[inline]
+		fn $name(&mut self, v: $val_type) -> Result<(), Error> {
+			Ok(self.writer.write_all(&$convfn(v))?)
+		}
+	}
+}
+
+macro_rules! reader_fn {
+	($name:ident, $val_type:ty, $val_size: expr, $convfn:ident) => {
+		#[inline]
+		fn $name(&mut self) -> Result<$val_type, Error> {
+			let mut buf = [0; $val_size];
+			self.reader.read_exact(&mut buf)?;
+			Ok($convfn(&buf))
+		}
+	}
+}
+
+use util::byte_utils::{be64_to_array, be32_to_array, be16_to_array, slice_to_be16, slice_to_be32, slice_to_be64};
+
+impl<W: Write> Writer<W> {
+	pub fn new(writer: W) -> Writer<W> {
+		return Writer { writer }
+	}
+	pub fn into_inner(self) -> W { self.writer }
+	pub fn get_ref(&self) -> &W { &self.writer }
+	writer_fn!(write_u64, u64, be64_to_array);
+	writer_fn!(write_u32, u32, be32_to_array);
+	writer_fn!(write_u16, u16, be16_to_array);
+	fn write_u8(&mut self, v: u8) -> Result<(), Error> { 
+		Ok(self.writer.write_all(&[v])?)
+	}
+	fn write_bool(&mut self, v: bool) -> Result<(), Error> {
+		Ok(self.writer.write_all(&[if v {1} else {0}])?)
+	}
+}
+
+impl<R: Read> Reader<R> {
+	pub fn new(reader: R) -> Reader<R> {
+		return Reader { reader }
+	}
+  pub fn into_inner(self) -> R { self.reader }
+	pub fn get_ref(&self) -> &R { &self.reader }
+
+	reader_fn!(read_u16, u16, 2, slice_to_be16);
+	reader_fn!(read_u32, u32, 4, slice_to_be32);
+	reader_fn!(read_u64, u64, 8, slice_to_be64);
+
+	fn read_u8(&mut self) -> Result<u8, Error> { 
+		let mut buf = [0; 1];
+		self.reader.read_exact(&mut buf)?;
+		Ok(buf[0])
+	}
+	fn read_bool(&mut self) -> Result<bool, Error> {
+		let mut buf = [0; 1];
+		self.reader.read_exact(&mut buf)?;
+		if buf[0] != 0 && buf[0] != 1 {
+			return Err(Error::InvalidValue);
+		}
+		Ok(buf[0] == 1)
+	}
+	fn read_fixed_size(&mut self, sz: usize) -> Result<Vec<u8>, Error> {
```

Comment:
Why not have the same API as reader.read_exact? Seems annyoing to force a vec allocation.

Reply:
Fixed